File: pages/refactorings/extract-method/01-order-total/After.scala
// Total of an order: the lines summed, less a percentage discount, plus tax.
object After:
  case class Line(unitPrice: Int, quantity: Int)
  case class Order(items: List[Line], discountPct: Int, taxPct: Int)

  def total(o: Order): Int =
    val net        = subtotal(o.items)
    val discounted = net - percent(o.discountPct, net)
    discounted + percent(o.taxPct, discounted)

  def subtotal(items: List[Line]): Int =
    items.map(l => l.unitPrice * l.quantity).sum

  def percent(pct: Int, amount: Int): Int =
    amount * pct / 100

File: pages/refactorings/extract-method/01-order-total/After.hs
-- Total of an order: the lines summed, less a percentage discount, plus tax.
module After where

data Line  = Line  { unitPrice :: Int, quantity :: Int }
data Order = Order { items :: [Line], discountPct :: Int, taxPct :: Int }

total :: Order -> Int
total o = discounted + percent (taxPct o) discounted
  where
    net        = subtotal (items o)
    discounted = net - percent (discountPct o) net

subtotal :: [Line] -> Int
subtotal ls = sum [unitPrice l * quantity l | l <- ls]

percent :: Int -> Int -> Int
percent pct amount = amount * pct `div` 100
